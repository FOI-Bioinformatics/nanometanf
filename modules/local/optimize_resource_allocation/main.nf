process OPTIMIZE_RESOURCE_ALLOCATION {
    tag "$meta.id"
    label 'process_single'
    publishDir "${params.outdir}/resource_analysis", mode: 'copy'

    input:
    tuple val(meta), path(predictions), path(system_metrics)
    val resource_config

    output:
    tuple val(meta), path("${meta.id}_optimal_allocation.json"), emit: allocations
    tuple val(meta), val(process_config), emit: process_configs
    tuple val(meta), path("${meta.id}_performance_metrics.json"), emit: performance_metrics
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    
    """
    #!/usr/bin/env python3
    
    import json
    import os
    import math
    from datetime import datetime
    from pathlib import Path
    
    # Load input data
    meta = ${groovy.json.JsonBuilder(meta).toString()}
    resource_config = ${groovy.json.JsonBuilder(resource_config).toString()}
    
    # Load predictions and system metrics
    with open('${predictions}', 'r') as f:
        predictions = json.load(f)
    
    with open('${system_metrics}', 'r') as f:
        system_metrics = json.load(f)
    
    def calculate_system_capacity():
        \"\"\"Calculate current system capacity and constraints\"\"\"
        cpu_info = system_metrics.get('cpu', {})
        memory_info = system_metrics.get('memory', {})
        current_load = system_metrics.get('current_load', {})
        
        # Calculate available resources considering current load
        total_cores = cpu_info.get('logical_cores', 4)
        total_memory_gb = memory_info.get('total_gb', 8)
        available_memory_gb = memory_info.get('available_gb', 4)
        
        cpu_utilization = current_load.get('cpu_utilization_percent', 0) / 100.0
        memory_utilization = current_load.get('memory_utilization_percent', 0) / 100.0
        
        # Apply safety factors
        safety_factor = resource_config.get('safety_factor', 0.8)
        
        capacity = {
            'max_cpu_cores': max(1, int(total_cores * safety_factor)),
            'max_memory_gb': max(2, available_memory_gb * safety_factor),
            'cpu_pressure': cpu_utilization > 0.75,
            'memory_pressure': memory_utilization > 0.8,
            'load_classification': current_load.get('load_classification', 'medium'),
            'available_cpu_cores': max(1, int(total_cores * (1 - cpu_utilization) * safety_factor)),
            'available_memory_gb': max(2, available_memory_gb * safety_factor)
        }
        
        return capacity
    
    def optimize_cpu_allocation(predicted_cores, system_capacity, priority):
        \"\"\"Optimize CPU allocation based on system constraints\"\"\"
        max_cores = system_capacity['max_cpu_cores']
        available_cores = system_capacity['available_cpu_cores']
        
        # Priority-based allocation
        priority_factors = {
            'high': 1.0,     # Get full requested resources
            'normal': 0.8,   # Get 80% of requested resources
            'low': 0.6       # Get 60% of requested resources
        }
        
        priority_factor = priority_factors.get(priority, 0.8)
        
        # Apply priority adjustment
        adjusted_cores = int(predicted_cores * priority_factor)
        
        # System constraint checks
        if system_capacity['cpu_pressure']:
            # Reduce allocation under high CPU pressure
            adjusted_cores = min(adjusted_cores, max(1, available_cores // 2))
        else:
            # Normal allocation up to available cores
            adjusted_cores = min(adjusted_cores, available_cores)
        
        # Ensure minimum and maximum bounds
        final_cores = max(1, min(adjusted_cores, max_cores))
        
        return {
            'allocated_cores': final_cores,
            'requested_cores': predicted_cores,
            'priority_adjustment': priority_factor,
            'constraint_limited': final_cores < predicted_cores,
            'allocation_efficiency': final_cores / predicted_cores if predicted_cores > 0 else 1.0
        }
    
    def optimize_memory_allocation(predicted_memory_gb, system_capacity, priority):
        \"\"\"Optimize memory allocation based on system constraints\"\"\"
        max_memory = system_capacity['max_memory_gb']
        available_memory = system_capacity['available_memory_gb']
        
        # Priority-based allocation
        priority_factors = {
            'high': 1.0,
            'normal': 0.85,
            'low': 0.7
        }
        
        priority_factor = priority_factors.get(priority, 0.85)
        
        # Apply priority adjustment
        adjusted_memory = predicted_memory_gb * priority_factor
        
        # System constraint checks
        if system_capacity['memory_pressure']:
            # Conservative allocation under memory pressure
            adjusted_memory = min(adjusted_memory, available_memory * 0.5)
        else:
            # Normal allocation up to available memory
            adjusted_memory = min(adjusted_memory, available_memory)
        
        # Ensure minimum and maximum bounds
        final_memory = max(2, min(adjusted_memory, max_memory))
        
        return {
            'allocated_memory_gb': round(final_memory, 1),
            'requested_memory_gb': predicted_memory_gb,
            'priority_adjustment': priority_factor,
            'constraint_limited': final_memory < predicted_memory_gb,
            'allocation_efficiency': final_memory / predicted_memory_gb if predicted_memory_gb > 0 else 1.0
        }
    
    def optimize_gpu_allocation(tool_context, system_capacity):
        \"\"\"Optimize GPU allocation based on tool requirements and availability\"\"\"
        gpu_info = system_metrics.get('gpu', {})
        tool_name = tool_context.get('tool_name', 'unknown')
        
        gpu_allocation = {
            'gpu_enabled': False,
            'gpu_devices': None,
            'gpu_memory_gb': 0,
            'acceleration_factor': 1.0
        }
        
        # Check if GPU is beneficial for this tool
        gpu_beneficial_tools = ['dorado_basecaller', 'guppy_basecaller']
        
        if tool_name in gpu_beneficial_tools and gpu_info.get('gpu_available'):
            nvidia_gpus = gpu_info.get('nvidia_gpus', [])
            apple_gpus = gpu_info.get('apple_gpus', [])
            
            if nvidia_gpus:
                # Use NVIDIA GPU
                gpu_allocation['gpu_enabled'] = True
                gpu_allocation['gpu_devices'] = 'cuda:all'
                gpu_allocation['gpu_memory_gb'] = sum(gpu['memory_total_mb'] for gpu in nvidia_gpus) / 1024
                gpu_allocation['acceleration_factor'] = 5.0  # 5x speedup estimate
            elif apple_gpus:
                # Use Apple Silicon GPU
                gpu_allocation['gpu_enabled'] = True
                gpu_allocation['gpu_devices'] = 'metal'
                gpu_allocation['gpu_memory_gb'] = 'shared'  # Apple Silicon shares system memory
                gpu_allocation['acceleration_factor'] = 3.0  # 3x speedup estimate
        
        return gpu_allocation
    
    def calculate_optimal_parallelization(predictions, optimized_resources, tool_context):
        \"\"\"Calculate optimal parallelization strategy\"\"\"
        tool_name = tool_context.get('tool_name', 'unknown')
        allocated_cores = optimized_resources['cpu']['allocated_cores']
        
        # Tool-specific parallelization strategies
        parallelization_profiles = {
            'dorado_basecaller': {
                'max_parallel_jobs': 1,  # GPU-bound, single job optimal
                'thread_scaling': 'linear',
                'batch_optimization': True
            },
            'kraken2': {
                'max_parallel_jobs': min(4, allocated_cores // 2),
                'thread_scaling': 'diminishing',
                'memory_per_job': 8  # GB
            },
            'filtlong': {
                'max_parallel_jobs': min(allocated_cores, 8),
                'thread_scaling': 'linear',
                'io_bound': True
            },
            'fastp': {
                'max_parallel_jobs': min(allocated_cores // 2, 4),
                'thread_scaling': 'linear',
                'thread_optimal': 4
            },
            'flye': {
                'max_parallel_jobs': 1,  # Single-threaded assembly
                'thread_scaling': 'internal',
                'cpu_intensive': True
            },
            'miniasm': {
                'max_parallel_jobs': 2,
                'thread_scaling': 'limited',
                'memory_intensive': True
            }
        }
        
        profile = parallelization_profiles.get(tool_name, parallelization_profiles.get('filtlong'))
        
        # Calculate optimal job configuration
        max_jobs = profile.get('max_parallel_jobs', 1)
        
        if profile.get('memory_per_job'):
            # Memory-limited parallelization
            memory_limited_jobs = int(optimized_resources['memory']['allocated_memory_gb'] / profile['memory_per_job'])
            max_jobs = min(max_jobs, memory_limited_jobs)
        
        threads_per_job = max(1, allocated_cores // max(max_jobs, 1))
        
        return {
            'parallel_jobs': max_jobs,
            'threads_per_job': threads_per_job,
            'total_threads': max_jobs * threads_per_job,
            'parallelization_strategy': profile.get('thread_scaling', 'linear'),
            'optimization_notes': profile
        }
    
    def generate_process_configuration(optimized_resources, parallelization, tool_context):
        \"\"\"Generate Nextflow process configuration\"\"\"
        cpu_config = optimized_resources['cpu']
        memory_config = optimized_resources['memory']
        gpu_config = optimized_resources['gpu']
        
        # Base process configuration
        process_config = {
            'cpus': cpu_config['allocated_cores'],
            'memory': f"{memory_config['allocated_memory_gb']}.GB",
            'time': f"{optimized_resources['runtime']['adjusted_time_hours']:.1f}h"
        }
        
        # Add GPU configuration if enabled
        if gpu_config['gpu_enabled']:
            if 'nvidia' in str(gpu_config['gpu_devices']):
                process_config['accelerator'] = [1, 'nvidia-tesla-v100']  # Generic NVIDIA config
            else:
                process_config['accelerator'] = [1, 'apple-silicon-gpu']  # Apple Silicon
        
        # Add tool-specific configurations
        tool_name = tool_context.get('tool_name', 'unknown')
        
        if tool_name == 'kraken2':
            # Kraken2 needs significant memory
            process_config['memory'] = f"{max(16, memory_config['allocated_memory_gb'])}.GB"
        elif tool_name in ['flye', 'miniasm']:
            # Assembly tools need extra time
            extended_time = optimized_resources['runtime']['adjusted_time_hours'] * 1.5
            process_config['time'] = f"{extended_time:.1f}h"
        
        # Add retry and error strategies
        process_config['errorStrategy'] = 'retry'
        process_config['maxRetries'] = 2
        
        return process_config
    
    def calculate_performance_metrics(predictions, optimized_resources, system_capacity):
        \"\"\"Calculate performance and efficiency metrics\"\"\"
        cpu_efficiency = optimized_resources['cpu']['allocation_efficiency']
        memory_efficiency = optimized_resources['memory']['allocation_efficiency']
        
        # Resource utilization scores
        cpu_utilization = optimized_resources['cpu']['allocated_cores'] / system_capacity['max_cpu_cores']
        memory_utilization = optimized_resources['memory']['allocated_memory_gb'] / system_capacity['max_memory_gb']
        
        # Performance predictions
        baseline_runtime = predictions['predictions']['runtime_estimates']['predicted_runtime_hours']
        
        # Apply acceleration factors
        gpu_factor = optimized_resources['gpu']['acceleration_factor']
        cpu_factor = optimized_resources['cpu']['allocated_cores'] / predictions['predictions']['cpu_requirements']['predicted_cores']
        
        adjusted_runtime = baseline_runtime / (gpu_factor * min(cpu_factor, 2.0))  # Cap CPU scaling benefit
        
        performance_metrics = {
            'resource_efficiency': {
                'cpu_efficiency': round(cpu_efficiency, 3),
                'memory_efficiency': round(memory_efficiency, 3),
                'overall_efficiency': round((cpu_efficiency + memory_efficiency) / 2, 3)
            },
            'system_utilization': {
                'cpu_utilization': round(cpu_utilization, 3),
                'memory_utilization': round(memory_utilization, 3),
                'resource_pressure': system_capacity['cpu_pressure'] or system_capacity['memory_pressure']
            },
            'performance_predictions': {
                'baseline_runtime_hours': round(baseline_runtime, 2),
                'optimized_runtime_hours': round(adjusted_runtime, 2),
                'speedup_factor': round(baseline_runtime / adjusted_runtime, 2) if adjusted_runtime > 0 else 1.0,
                'gpu_acceleration': optimized_resources['gpu']['gpu_enabled']
            },
            'optimization_summary': {
                'cpu_constraint_limited': optimized_resources['cpu']['constraint_limited'],
                'memory_constraint_limited': optimized_resources['memory']['constraint_limited'],
                'system_load_impact': system_capacity['load_classification'],
                'allocation_confidence': predictions['confidence_metrics']['confidence_level']
            }
        }
        
        return performance_metrics
    
    # Main optimization logic
    print(f"Optimizing resource allocation for {meta['id']}")
    
    # Extract key information
    tool_context = predictions.get('tool_context', {})
    cpu_prediction = predictions['predictions']['cpu_requirements']['predicted_cores']
    memory_prediction = predictions['predictions']['memory_requirements']['predicted_memory_gb']
    priority = predictions['recommendations'].get('priority_level', 'normal')
    
    print(f"Tool: {tool_context.get('tool_name', 'unknown')}")
    print(f"Predicted requirements - CPU: {cpu_prediction} cores, Memory: {memory_prediction:.1f} GB")
    print(f"Priority: {priority}")
    
    # Calculate system capacity
    system_capacity = calculate_system_capacity()
    print(f"System capacity - CPU: {system_capacity['available_cpu_cores']} cores, Memory: {system_capacity['available_memory_gb']:.1f} GB")
    
    # Optimize resource allocations
    cpu_optimization = optimize_cpu_allocation(cpu_prediction, system_capacity, priority)
    memory_optimization = optimize_memory_allocation(memory_prediction, system_capacity, priority)
    gpu_optimization = optimize_gpu_allocation(tool_context, system_capacity)
    
    # Calculate runtime adjustments
    runtime_optimization = {
        'baseline_time_hours': predictions['predictions']['runtime_estimates']['predicted_runtime_hours'],
        'adjusted_time_hours': predictions['predictions']['runtime_estimates']['predicted_runtime_hours'] / gpu_optimization['acceleration_factor']
    }
    
    # Combine optimized resources
    optimized_resources = {
        'cpu': cpu_optimization,
        'memory': memory_optimization,
        'gpu': gpu_optimization,
        'runtime': runtime_optimization
    }
    
    # Calculate optimal parallelization
    parallelization = calculate_optimal_parallelization(predictions, optimized_resources, tool_context)
    
    # Generate process configuration
    process_config = generate_process_configuration(optimized_resources, parallelization, tool_context)
    
    # Calculate performance metrics
    performance_metrics = calculate_performance_metrics(predictions, optimized_resources, system_capacity)
    
    # Compile final allocation
    optimal_allocation = {
        'sample_id': meta['id'],
        'optimization_timestamp': datetime.now().isoformat(),
        'tool_context': tool_context,
        'system_capacity': system_capacity,
        'original_predictions': {
            'cpu_cores': cpu_prediction,
            'memory_gb': memory_prediction,
            'runtime_hours': predictions['predictions']['runtime_estimates']['predicted_runtime_hours']
        },
        'optimized_allocation': {
            'cpu_cores': cpu_optimization['allocated_cores'],
            'memory_gb': memory_optimization['allocated_memory_gb'],
            'gpu_enabled': gpu_optimization['gpu_enabled'],
            'estimated_runtime_hours': runtime_optimization['adjusted_time_hours']
        },
        'parallelization_strategy': parallelization,
        'optimization_details': optimized_resources,
        'performance_predictions': performance_metrics['performance_predictions'],
        'nextflow_process_config': process_config
    }
    
    # Save optimal allocation
    output_file = f"{meta['id']}_optimal_allocation.json"
    with open(output_file, 'w') as f:
        json.dump(optimal_allocation, f, indent=2)
    
    # Save performance metrics
    performance_file = f"{meta['id']}_performance_metrics.json"
    with open(performance_file, 'w') as f:
        json.dump(performance_metrics, f, indent=2)
    
    print(f"Optimization complete for {meta['id']}:")
    print(f"  Allocated CPU cores: {cpu_optimization['allocated_cores']} (requested: {cpu_prediction})")
    print(f"  Allocated memory: {memory_optimization['allocated_memory_gb']:.1f} GB (requested: {memory_prediction:.1f} GB)")
    print(f"  GPU acceleration: {'enabled' if gpu_optimization['gpu_enabled'] else 'disabled'}")
    print(f"  Estimated runtime: {runtime_optimization['adjusted_time_hours']:.1f} hours")
    print(f"  Resource efficiency: {performance_metrics['resource_efficiency']['overall_efficiency']:.2f}")
    
    # Generate versions file
    with open('versions.yml', 'w') as f:
        f.write('''\"${task.process}\":
    python: \"3.9\"
    resource_optimization: \"1.0\"''')
    """

    stub:
    """
    echo '{"sample_id": "${meta.id}", "stub": true, "optimized_allocation": {"cpu_cores": 4, "memory_gb": 8}}' > ${meta.id}_optimal_allocation.json
    echo '{"resource_efficiency": {"overall_efficiency": 0.8}}' > ${meta.id}_performance_metrics.json
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: "3.9"
        resource_optimization: "1.0"
    END_VERSIONS
    """
}