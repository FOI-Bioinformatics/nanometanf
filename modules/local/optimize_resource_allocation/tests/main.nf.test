nextflow_process {

    name "Test OPTIMIZE_RESOURCE_ALLOCATION"
    script "../main.nf"
    process "OPTIMIZE_RESOURCE_ALLOCATION"

    tag "modules"
    tag "modules_local"
    tag "optimize_resource_allocation"

    test("Should optimize basic resource allocation") {

        tag "basic_optimization"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'basic_optimization_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/predictions/balanced_prediction.json'),
                    file('$projectDir/tests/fixtures/system_metrics/normal_load.json')
                ]
                input[1] = [
                    safety_factor: 0.8,
                    enable_gpu_optimization: false
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle high priority workload optimization") {

        tag "high_priority"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'high_priority_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/predictions/high_throughput_prediction.json'),
                    file('$projectDir/tests/fixtures/system_metrics/high_load.json')
                ]
                input[1] = [
                    safety_factor: 0.9,
                    priority_boost: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should optimize GPU-enabled workload") {

        tag "gpu_optimization"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'gpu_optimization_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/predictions/balanced_prediction.json'),
                    file('$projectDir/tests/fixtures/system_metrics/gpu_available.json')
                ]
                input[1] = [
                    safety_factor: 0.8,
                    enable_gpu_optimization: true,
                    gpu_acceleration_strategy: "aggressive"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle system under resource pressure") {

        tag "resource_pressure"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'pressure_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/predictions/conservative_prediction.json'),
                    file('$projectDir/tests/fixtures/system_metrics/high_load.json')
                ]
                input[1] = [
                    safety_factor: 0.6,
                    conservative_mode: true,
                    pressure_management: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should optimize for low priority workload") {

        tag "low_priority"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'low_priority_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/predictions/balanced_prediction.json'),
                    file('$projectDir/tests/fixtures/system_metrics/normal_load.json')
                ]
                input[1] = [
                    safety_factor: 0.7,
                    background_mode: true,
                    shared_system: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle Apple Silicon GPU optimization") {

        tag "apple_silicon"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'apple_gpu_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/predictions/balanced_prediction.json'),
                    file('$projectDir/tests/fixtures/system_metrics/gpu_available.json')
                ]
                input[1] = [
                    safety_factor: 0.85,
                    enable_gpu_optimization: true,
                    apple_silicon_optimizations: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("Should handle invalid predictions gracefully") {

        tag "edge_cases"
        tag "snapshot"

        when {
            process {
                """
                input[0] = [
                    [
                        id: 'invalid_input_test',
                        single_end: true
                    ],
                    file('$projectDir/tests/fixtures/edge_cases/extreme_predictions.json'),
                    file('$projectDir/tests/fixtures/system_metrics/normal_load.json')
                ]
                input[1] = [
                    safety_factor: 0.8,
                    handle_invalid_input: true,
                    use_fallback_defaults: true
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

}