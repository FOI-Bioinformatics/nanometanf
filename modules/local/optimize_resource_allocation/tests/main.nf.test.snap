{
    "Should handle Apple Silicon GPU optimization": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "apple_gpu_test",
                            "single_end": true
                        },
                        "apple_gpu_test_optimal_allocation.json:md5,66c738d6ed43ecacf87ae017338d14d4"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "apple_gpu_test",
                            "single_end": true
                        },
                        "apple_gpu_test_performance_metrics.json:md5,051991da19c252e0c163b25c05f2190c"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ],
                "allocations": [
                    [
                        {
                            "id": "apple_gpu_test",
                            "single_end": true
                        },
                        "apple_gpu_test_optimal_allocation.json:md5,66c738d6ed43ecacf87ae017338d14d4"
                    ]
                ],
                "performance_metrics": [
                    [
                        {
                            "id": "apple_gpu_test",
                            "single_end": true
                        },
                        "apple_gpu_test_performance_metrics.json:md5,051991da19c252e0c163b25c05f2190c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:26:51.514304"
    },
    "Should optimize for low priority workload": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "low_priority_test",
                            "single_end": true
                        },
                        "low_priority_test_optimal_allocation.json:md5,3a554c36d71da665f1a69dc182511d5d"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "low_priority_test",
                            "single_end": true
                        },
                        "low_priority_test_performance_metrics.json:md5,5828426485d773efc17b9218ec204947"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ],
                "allocations": [
                    [
                        {
                            "id": "low_priority_test",
                            "single_end": true
                        },
                        "low_priority_test_optimal_allocation.json:md5,3a554c36d71da665f1a69dc182511d5d"
                    ]
                ],
                "performance_metrics": [
                    [
                        {
                            "id": "low_priority_test",
                            "single_end": true
                        },
                        "low_priority_test_performance_metrics.json:md5,5828426485d773efc17b9218ec204947"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:26:47.789046"
    },
    "Should optimize GPU-enabled workload": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "gpu_optimization_test",
                            "single_end": true
                        },
                        "gpu_optimization_test_optimal_allocation.json:md5,746cbcdd2856d13e3bb0474e282b3496"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "gpu_optimization_test",
                            "single_end": true
                        },
                        "gpu_optimization_test_performance_metrics.json:md5,26e6fab7b8ca0ca5f3a580da464ff08e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ],
                "allocations": [
                    [
                        {
                            "id": "gpu_optimization_test",
                            "single_end": true
                        },
                        "gpu_optimization_test_optimal_allocation.json:md5,746cbcdd2856d13e3bb0474e282b3496"
                    ]
                ],
                "performance_metrics": [
                    [
                        {
                            "id": "gpu_optimization_test",
                            "single_end": true
                        },
                        "gpu_optimization_test_performance_metrics.json:md5,26e6fab7b8ca0ca5f3a580da464ff08e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:26:39.543837"
    },
    "Should handle high priority workload optimization": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "high_priority_test",
                            "single_end": true
                        },
                        "high_priority_test_optimal_allocation.json:md5,3fa38a1fe86bc379dd37026b9439f607"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "high_priority_test",
                            "single_end": true
                        },
                        "high_priority_test_performance_metrics.json:md5,59019368cefb4120b3507e4f44208c8f"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ],
                "allocations": [
                    [
                        {
                            "id": "high_priority_test",
                            "single_end": true
                        },
                        "high_priority_test_optimal_allocation.json:md5,3fa38a1fe86bc379dd37026b9439f607"
                    ]
                ],
                "performance_metrics": [
                    [
                        {
                            "id": "high_priority_test",
                            "single_end": true
                        },
                        "high_priority_test_performance_metrics.json:md5,59019368cefb4120b3507e4f44208c8f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:26:35.693796"
    },
    "Should optimize basic resource allocation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "basic_optimization_test",
                            "single_end": true
                        },
                        "basic_optimization_test_optimal_allocation.json:md5,1f99903033c7ae8b6aaadd296a53fce0"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "basic_optimization_test",
                            "single_end": true
                        },
                        "basic_optimization_test_performance_metrics.json:md5,26e6fab7b8ca0ca5f3a580da464ff08e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ],
                "allocations": [
                    [
                        {
                            "id": "basic_optimization_test",
                            "single_end": true
                        },
                        "basic_optimization_test_optimal_allocation.json:md5,1f99903033c7ae8b6aaadd296a53fce0"
                    ]
                ],
                "performance_metrics": [
                    [
                        {
                            "id": "basic_optimization_test",
                            "single_end": true
                        },
                        "basic_optimization_test_performance_metrics.json:md5,26e6fab7b8ca0ca5f3a580da464ff08e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:26:32.118965"
    },
    "Should handle system under resource pressure": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "pressure_test",
                            "single_end": true
                        },
                        "pressure_test_optimal_allocation.json:md5,d6e6d5a5547a8dfac7783506ad10ab7b"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "pressure_test",
                            "single_end": true
                        },
                        "pressure_test_performance_metrics.json:md5,33f98cf3e8a95cb876dc3e1867c7e451"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ],
                "allocations": [
                    [
                        {
                            "id": "pressure_test",
                            "single_end": true
                        },
                        "pressure_test_optimal_allocation.json:md5,d6e6d5a5547a8dfac7783506ad10ab7b"
                    ]
                ],
                "performance_metrics": [
                    [
                        {
                            "id": "pressure_test",
                            "single_end": true
                        },
                        "pressure_test_performance_metrics.json:md5,33f98cf3e8a95cb876dc3e1867c7e451"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:26:43.954747"
    },
    "Should handle invalid predictions gracefully": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "invalid_input_test",
                            "single_end": true
                        },
                        "invalid_input_test_optimal_allocation.json:md5,5a75359693d01b3b41685a9a701d5972"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "invalid_input_test",
                            "single_end": true
                        },
                        "invalid_input_test_performance_metrics.json:md5,ec19ea044a3678158ef30f830e2cf20c"
                    ]
                ],
                "2": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ],
                "allocations": [
                    [
                        {
                            "id": "invalid_input_test",
                            "single_end": true
                        },
                        "invalid_input_test_optimal_allocation.json:md5,5a75359693d01b3b41685a9a701d5972"
                    ]
                ],
                "performance_metrics": [
                    [
                        {
                            "id": "invalid_input_test",
                            "single_end": true
                        },
                        "invalid_input_test_performance_metrics.json:md5,ec19ea044a3678158ef30f830e2cf20c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2a2e97bf7e41c0b1f7feb5bc82564f3f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T09:26:55.02653"
    }
}